### PR TITLE
Update coroutines example to be the new newsreader example

### PR DESCRIPTION
**Pull request by edualonso**
Updated coroutines example to be the new newsreader example. All data is accessed and consumed asynchronously using Kotlin Flows.

- The main view is a recyclerview with a pull-to-refresh mechanism and a dropdown spinner to select a section.
- Stories are valid for 30 minutes. After said period the next load will refresh the current section.
- The details view is just a few textviews. A news story will be marked as "read" after 2 seconds in the details view.
- Views are inflated using Android DataBinding and UI updates are handled via LiveData.
- Individual stories for the details view are fetched directly from Realm and not Store since we already have them in disk.
- The project uses Dropbox Store (https://github.com/dropbox/Store) for fetching/accessing the stories from the main view.
- Fetching data from the NYT API is done using Retrofit.
- A custom FlowFactory is showcased due to Store's logic when processing results in its data reader. Store does not consider an empty list as no results (`toFlow` and `asFlowable` usually emit an empty `RealmResults` list when calling `findAllAsync`), therefore it will treat it as "results not found", triggering the API fetcher and therefore rendering the repository useless. In this custom flow factory implementation the current results/objects are not emitted right away, which does not confuse Store into thinking it has to refresh data from the API.
- There are data models for network and domain layers. Realm models are used as domain objects.

### CODE REVIEW

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/di/DependencyGraph.kt
Comment (cmelchior) [inline examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/di/DependencyGraph.kt:35]: Maybe just add a comment that this is our own DI solution for the sake of keeping the example simple?

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/details/DetailsViewModel.kt
Comment (cmelchior) [inline examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/details/DetailsViewModel.kt:74]: Is this check needed? I thought that delay would abort in a clean way if the coroutine is closed?

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/main/MainFragment.kt
Comment (cmelchior) [inline examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/main/MainFragment.kt:155]: Any reason we need both `ErrorException` and `ErrorMessage` as states? Doesn't that just complicate things since we are just reading the error message from the exception anyway?

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/util/NewsReaderFlowFactory.kt
Comment (cmelchior) [inline examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/util/NewsReaderFlowFactory.kt:33]: Wouldn't a much cleaner solution be to just call the super methods and then drop or filter out the first element emitted?

Copying the entire internal implementation seems very error-prone in terms of maintaining it?

**Review by cmelchior — APPROVED** [review]
Really nice 🥇 (sorry for being late with the review)

Most of the comments I left were around adding a little more documentation and a few code comments.

 Two "major" comments:
- Shouldn't we just delete the old newsReaderExample? I'm not sure it provides much value anymore once this is merged?
- You have an ObjectStore commit as part of this PR. That looks wrong?

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/details/DetailsViewModel.kt
Comment (edualonso) [reply examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/details/DetailsViewModel.kt:74]: That's correct, I forgot the flow is launched used `viewModelScope`.

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/main/MainFragment.kt
Comment (edualonso) [reply examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/ui/main/MainFragment.kt:155]: IIRC this is because Store can emit either an "error exception" or an "error message". I decided to leave it the same way for consistency.

File: examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/util/NewsReaderFlowFactory.kt
Comment (edualonso) [reply examples/coroutinesExample/src/main/java/io/realm/examples/coroutinesexample/util/NewsReaderFlowFactory.kt:33]: Well spotted. I also thought about doing that but when I did this the updated factory (with the latest changes and such) wasn't in the source branch for this branch, so I copied it entirely from the flow factory branch I was simultaneously working on as temporary solution. It seems I forgot to clean it up 🙊

**Comment by edualonso** [comment]
I've addressed all comments and updated the OS pointer and will merge this since I got a green light 🚀

**Pull request merged**
